lib/translator/translation.cc

```diff
@@ -69,39 +70,69 @@ BeginResponse Translation::Begin(Span<const uint8_t> scsi_cmd,
   return response;
 }
 
-ApiStatus Translation::Complete(Span<const nvme::GenericQueueEntryCpl> cpl_data,
-                                Span<uint8_t> buffer) {
+CompleteResponse Translation::Complete(
+    Span<const nvme::GenericQueueEntryCpl> cpl_data, Span<uint8_t> buffer_in,
+    Span<uint8_t> sense_buffer) {
+  CompleteResponse resp = {};
   if (pipeline_status_ == StatusCode::kUninitialized) {
     DebugLog("Invalid use of API: Complete called before Begin");
-    return ApiStatus::kFailure;
+    resp.status = ApiStatus::kFailure;
+    return resp;
   }
+
+  if (cpl_data.size() != nvme_cmd_count_) {
+    DebugLog(
+        "Invalid use of API, completion count %u does not equal command count "
+        "%u",
+        cpl_data.size(), nvme_cmd_count_);
+    resp.status = ApiStatus::kFailure;
+    return resp;
+  }
+
   if (pipeline_status_ == StatusCode::kFailure) {
-    // TODO fill buffer with SCSI CHECK CONDITION response
+    ScsiStatus scsi_status = {
+        .status = scsi::Status::kCheckCondition,
+        .sense_key = scsi::SenseKey::kIllegalRequest,
+        .asc = scsi::AdditionalSenseCode::kInvalidFieldInCdb,
+        .ascq = scsi::AdditionalSenseCodeQualifier::kNoAdditionalSenseInfo};
+    FillSenseBuffer(sense_buffer, scsi_status);
     AbortPipeline();
-    return ApiStatus::kSuccess;
+    resp.status = ApiStatus::kSuccess;
+    resp.scsi_status = scsi_status.status;
+    return resp;
+  }
+
+  // Verify NVMe commands completed successfully. If not translate error.
+  for (uint32_t i = 0; i < cpl_data.size(); ++i) {
+    const nvme::GenericQueueEntryCpl& cpl_entry = cpl_data[i];
+    const nvme::CplStatus& cpl_status = cpl_entry.cpl_status;
+    ScsiStatus scsi_status = StatusToScsi(cpl_status.sct, cpl_status.sc);
+    if (scsi_status.status != scsi::Status::kGood) {
+      FillSenseBuffer(sense_buffer, scsi_status);
+      resp.status = ApiStatus::kSuccess;
+      resp.scsi_status = scsi_status.status;
+      return resp;
+    }
   }
 
   // Switch cases should not return
-  ApiStatus ret;
+  resp.status = ApiStatus::kSuccess;
```

Comment:
on default switch case we are successful, is this correct behavior?

Reply:
Technically we should never have a default switch case in this section of the library because the switch statement is validated in begin, which is enforced to be called before complete. Any default cases would imply our library is missing a switch that is in begin. We can log an error in this case